CA-197067: DB upgrade rule: has_vendor_device=false

thomassa opened the pull request:
This is needed because the default value for VM.has_vendor_device
defined in the datamodel is used in two places:
* When upgrading the database from a version that lacks the field
  (where we want it false)
* When VM.create is called without specifying the field
  (where we want it true).

Therefore this new database upgrade rule sets has_vendor_device=false
for each VM, if the upgrade is from a version that lacks the field
(determined by checking whether it is a pre-dundee version).

thomassa commented:
This builds fine; I'm waiting for jobs to run to test whether it fixes the upgrade problem.
In the mean time, here's the change for review (of the approach taken as much as the new code itself).

robhoes commented on ocaml/xapi/xapi_db_upgrade.ml:
```diff
@@ -497,6 +506,7 @@ let rules = [
 	populate_pgpu_vgpu_types;
 	set_vgpu_types;
 	add_default_pif_properties;
+    default_has_vendor_device_false;
```
The patch looks good. Only the indentation here looks wrong.

thomassa commented:
It would be more satisfying and perhaps long-term safer to change the datamodel default to `false`, and special-case it to `true` in `VM.create` only. Not yet sure how viable that is though.

thomassa commented:
Fixing has_vendor_device on DB upgrade is not enough: we need to do it on VM import as well, if the VM is an old one that lacks the field.

thomassa commented:
Now I have fixed the indentation and added a commit to handle import of VMs that lack the `has_vendor_device` field.

It still might be better to use `false

thomassa commented:
Travis failure is unrelated:
```
[ERROR] Bad checksum for
        /home/travis/.opam/packages.dev/ssl.0.5.2/ocaml-ssl-0.5.2.tar.gz:
          - 404f71d33885c985a8ff579996a5cda8 [expected result]
          - 0103b909e19ca9c6497a7ae696c16480 [actual result]
        This may be fixed by running `opam update`.

( ...snip ...)

[ERROR] The sources of the following couldn't be obtained, aborting:
          - ssl.0.5.2
        (This may be fixed by running 'opam update')
'unset TESTS; OPAMYES=1 export OPAMYES; set -ue; opam install xapi --deps-only' exited 66. Terminating with 66
The command "bash -ex .travis-opam.sh" exited with 66.
Done. Your build exited with 1.
```
...so the test should be re-run once the problem with that ssl package has been fixed.

robhoes commented:
The import patch looks good to me as well :shipit:

thomassa commented:
Merging: this has passed BVT, and I have tested that both database-upgrade and import from older XenServer versions produce VMs with has_vendor_device=false.
